Sources/NIO/Codec.swift

```diff
@@ -674,3 +674,118 @@ extension MessageToByteHandler {
+        typealias Out = Decoder.InboundOut
+
+        func verifySimple(channel: EmbeddedChannel) throws {
+            for (input, expectedOutputs) in inputOutputPairs.shuffled() {
+                try channel.writeInbound(input)
+                for expectedOutput in expectedOutputs {
+                    let actualOutput = try channel.readInbound(as: Decoder.InboundOut.self)
+                    guard actualOutput == expectedOutput else {
+                        throw ByteToMessageDecoderVerificationError.wrongDecodingResult(actual: actualOutput,
+                                                                                        expected: expectedOutput)
+                    }
+                }
+                let actualExtraOutput = try channel.readInbound(as: Decoder.InboundOut.self)
+                guard actualExtraOutput == nil else {
+                    throw ByteToMessageDecoderVerificationError.unexpectedDecodingResult(actualExtraOutput)
+                }
+            }
+        }
+
+        func verifyDripFeed(channel: EmbeddedChannel) throws {
+            for _ in 0..<10 {
+                for (input, expectedOutputs) in inputOutputPairs.shuffled() {
+                    for c in input.readableBytesView {
+                        var buffer = channel.allocator.buffer(capacity: 12)
+                        buffer.writeString("BEFORE")
+                        buffer.writeInteger(c)
+                        buffer.writeString("AFTER")
+                        buffer.moveReaderIndex(forwardBy: 6)
+                        buffer.moveWriterIndex(to: buffer.readerIndex + 1)
+                        try channel.writeInbound(buffer)
+                    }
+                    for expectedOutput in expectedOutputs {
+                        let actualOutput = try channel.readInbound(as: Decoder.InboundOut.self)
+                        guard actualOutput == expectedOutput else {
+                            throw ByteToMessageDecoderVerificationError.wrongDecodingResult(actual: actualOutput,
+                                                                                            expected: expectedOutput)
+                        }
+                    }
+                    let actualExtraOutput = try channel.readInbound(as: Decoder.InboundOut.self)
+                    guard actualExtraOutput == nil else {
+                        throw ByteToMessageDecoderVerificationError.unexpectedDecodingResult(actualExtraOutput)
+                    }
+                }
+            }
+        }
+
+        func verifyManyAtOnce(channel: EmbeddedChannel) throws {
+            var overallBuffer = channel.allocator.buffer(capacity: 1024)
+            var overallExpecteds: [Out] = []
+
+            for _ in 0..<10 {
+                for (var input, expectedOutputs) in inputOutputPairs.shuffled() {
```

Comment:
It’s not necessarily the case that you can assume that you can shuffle the inputs and get the same result: for example, the HTTPDecoder cannot tolerate that.

Reply:
@Lukasa In general, you're right. B2MDs that retain state across one message can't be used like this. I think this is still useful as is to verify B2MDs that don't do that. We could offer some form of configuration but when I thought about it, it quickly became too complicated. For HTTPDecoder this means that this B2MD verifier will only work for keep-alive messages where I can shuffle.

Another issue is that some B2MDs (like HTTPDecoder) generate different outputs depending on whether the input came in a large chunk or was drip-fed. HTTP bodies for example are just handed through by HTTPDecoder in whatever shape they arrived. Hence, I'm using 1 byte bodies in the tests here.

I should probably make this more clear in the documentation but this is meant as a test for quite straightforward B2MDs, the ones that don't retain any state across messages and I think that's quite useful. This isn't supposed to replace the regular unit tests but it's supposed to give you a baseline. And most B2MDs (including HTTPDecoder) can be used in this way.

Reply:
Ok, I can get behind accepting this as a limitation for now.